jobserver/releases.py

```diff
@@ -125,38 +164,60 @@ def handle_file_upload(release, backend, user, upload, filename, **kwargs):
     # TODO: enforce a max upload size?
     data = upload.read()
     calculated_hash = hashlib.sha256(data).hexdigest()
-    check_not_already_uploaded(filename, calculated_hash, backend)
+    relative_path, absolute_path = _build_paths(release, filename, data)
 
-    # We group the directory structure by workspace for 2 reasons:
-    # 1. avoid dumping everything in one big directory.
-    # 2. much easier for human operators to navigate on disk if needed.
-    relative_path = (
-        Path(release.workspace.name) / "releases" / str(release.id) / filename
-    )
-    absolute_path = absolute_file_path(relative_path)
-    absolute_path.parent.mkdir(parents=True, exist_ok=True)
-    absolute_path.write_bytes(data)
-
-    mtime = datetime.fromtimestamp(absolute_path.stat().st_mtime, tz=timezone.utc)
-
-    try:
-        rfile = ReleaseFile.objects.create(
-            release=release,
-            workspace=release.workspace,
-            created_by=user,
-            name=filename,
-            path=str(relative_path),
-            filehash=calculated_hash,
-            mtime=mtime,
-            size=absolute_path.stat().st_size,
-            **kwargs,
+    # Check if this filename for this release has been uploaded before.
+    existing = ReleaseFile.objects.filter(
+        release=release,
+        release__backend=backend,
+        name=filename,
+    ).first()
+
+    if not existing:  # old flow
+        absolute_path.write_bytes(data)
+
+        mtime = datetime.fromtimestamp(absolute_path.stat().st_mtime, tz=timezone.utc)
+        size = absolute_path.stat().st_size
+
+        try:
+            return ReleaseFile.objects.create(
+                release=release,
+                workspace=release.workspace,
+                created_by=user,
+                uploaded_at=now(),
+                name=filename,
+                path=str(relative_path),
+                filehash=calculated_hash,
+                mtime=mtime,
+                size=size,
+                **kwargs,
+            )
+        except Exception:
+            # something went wrong, clean up file, they will need to reupload
+            absolute_path.unlink(missing_ok=True)
+            raise
+
+    # New flow
+
+    # _is_ on disk
+    if existing.absolute_path().exists():
+        # existence of a ReleaseFile isn't an error case now but a file-on-disk
+        # having already been uploaded is
+        raise ReleaseFileAlreadyExists(
+            f"This version of '{filename}' has already been uploaded from backend '{backend.slug}'"
```

Comment:
I wonder if this should be a 4xx rather than a 5xx?

The client has done something wrong here

Reply:
I think that's what it's doing, we're catching that exception in the API and raising a ValidationError from it which I would expect to give a 4xx?

Reply:
ok, great